### PR TITLE
DAOS-2768 control: Find daos_io_server via readlink

### PR DESCRIPTION
**Pull request by mjmac**
If daos_io_server can't be located via $PATH, fall back to
checking to see if it's adjacent to daos_server via readlink
on /proc/self/exe.

### CODE REVIEW

File: src/control/server/iosrv_test.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"testing"
+)
+
+func TestFindBinaryInPath(t *testing.T) {
```
Comment (kjacque) [inline src/control/server/iosrv_test.go]: Just a philosophical quibble, but I'm not wild about having to manipulate real filesystems and environment variables in unit tests. I'd prefer to move toward using interfaces and dependency injection in the Go code to make it easy to mock these kinds of things.

Comment (mjmac) [reply src/control/server/iosrv_test.go]: This kind of testing is how the stdlib tests itself... How would we test this using interfaces?

**Comment by mjmac** [comment]
> 
> 
> LGTM. That said, was there a reason why the daos_io_server would be moved from the default location? Do we have to look for other binaries that are exec'd to make sure we apply the same behavior to be consistent?

Not sure what the motivation was for opening the ticket, but in my own experience I have noticed that if I start daos_server with sudo and I forget to set the $PATH appropriately it will fail to start daos_io_server. This patch would fix that problem.

For the other question: Yes, this technique would allow us to find other binaries that aren't in the $PATH at runtime.

File: src/control/server/iosrv_test.go (hunk 1 above)
Comment (kjacque) [reply src/control/server/iosrv_test.go]: The strategy I like is creating a thin interface layer over the external packages. The default implementation just calls the real function, but a mock version can be injected specifically for the tests. In this case I could see calls like os.Readlink and os.Stat being mocked. Allows for easy testing of various failure paths.

**Pull request merged**